wetterdienst/core/scalar/interpolate.py

```diff
@@ -0,0 +1,254 @@
+    else:
+        param_data.finished = True
+
+
+def gain_of_value_pairs(old_values: pd.DataFrame, new_values: pd.Series) -> float:
+    old_score = old_values.apply(lambda row: row.dropna().size >= 4).sum()  # 5: dates plus 4 values
+    old_values[new_values.name] = new_values.values  # Add new column
+    new_score = old_values.apply(lambda row: row.dropna().size >= 4).sum()  # 5: dates plus 4 values
+    return new_score / old_score - 1
+
+
+def calculate_interpolation(utm_x: float, utm_y: float, stations_dict: dict, param_dict: dict) -> pd.DataFrame:
+    columns = [
+        Columns.DATE.value,
+        Columns.PARAMETER.value,
+        Columns.VALUE.value,
+        Columns.DISTANCE_MEAN.value,
+        Columns.STATION_IDS.value,
+    ]
+    param_df_list = [pd.DataFrame(columns=columns)]
+    valid_station_groups = get_valid_station_groups(stations_dict, utm_x, utm_y)
+
+    for parameter, param_data in param_dict.items():
+        param_df = pd.DataFrame(columns=columns)
+        param_df[columns[1:]] = param_data.values.apply(
+            lambda row, param=parameter: apply_interpolation(
+                row, stations_dict, valid_station_groups, param, utm_x, utm_y
+            ),
+            axis=1,
+            result_type="expand",
+        )
+        param_df[Columns.DATE.value] = param_data.values.index
+        param_df_list.append(param_df)
+
+    return pd.concat(param_df_list).sort_values(by=[Columns.DATE.value, Columns.PARAMETER.value]).reset_index(drop=True)
+
+
+def get_valid_station_groups(stations_dict: dict, utm_x: float, utm_y: float):
+    point = Point(utm_x, utm_y)
+
+    valid_groups = Queue()
+    # get all combinations of 4 stations
+    for station_group in combinations(stations_dict.keys(), 4):
+        coords = [(stations_dict[s][1], stations_dict[s][0]) for s in station_group]
+        pol = Polygon(coords)
+        if pol.contains(point):
+            valid_groups.put(station_group)
+
+    return valid_groups
+
+
+@lru_cache
+def get_station_group_ids(valid_station_groups: Queue, vals_index: frozenset) -> list:
+    for item in valid_station_groups.queue:
+        if set(item).issubset(vals_index):
+            return list(item)
+    return []
+
+
+def apply_interpolation(row, stations_dict: dict, valid_station_groups, parameter, utm_x: float, utm_y: float):
+    vals_state = ~pd.isna(row.values)
+    vals = row[vals_state].astype(float)
+
+    station_group_ids = get_station_group_ids(valid_station_groups, frozenset(vals.index))
+
+    if station_group_ids:
+        vals = vals[station_group_ids]
+    else:
+        vals = None
+
+    value = np.nan
+    distance_mean = np.nan
+
+    if vals is None or vals.size < 4:
+        return parameter, value, distance_mean, station_group_ids
+
+    xs, ys, distances = map(list, zip(*[stations_dict[station_id] for station_id in station_group_ids]))
+    distance_mean = sum(distances) / len(distances)
+
+    f = interpolate.interp2d(ys, xs, vals, kind="linear")
```

Comment:
Shouldn't this rather be 
```suggestion
    f = interpolate.interp2d(xs, ys, vals, kind="linear")
```
?